src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,146 @@
+        }
+        if (index == size) {
+            linkLast(value);
+        } else if (isEmpty()) {
+            head = tail = new Node<>(null, value, null);
+
+        } else if (index == 0) {
+            Node<T> newNode = new Node<>(null, value, head);
+            head = newNode;
+            newNode.next.prev = newNode;
+        } else {
+            linkBefore(value, getNodeByIndex(index));
+        }
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        list.forEach(this::add);
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndexInBounds(index);
+        return getNodeByIndex(index).value;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkIndexInBounds(index);
+        Node<T> node = getNodeByIndex(index);
+        T oldValue = node.value;
+        node.value = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndexInBounds(index);
+        Node<T> node = getNodeByIndex(index);
+        if (node.prev != null) {
+            node.prev.next = node.next;
+        } else {
+            head = node.next;
+        }
+        if (node.next != null) {
+            node.next.prev = node.prev;
+        } else {
+            tail = node.prev;
+        }
+        size--;
+        return node.value;
     }
 
     @Override
     public boolean remove(T object) {
+        for (int i = 0; i < size; i++) {
+            if (getNodeByIndex(i).value == object || getNodeByIndex(i).value != null
+                    && getNodeByIndex(i).value.equals(object)) {
+                remove(i);
+                return true;
+            }
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private Node<T> getNodeByIndex(int index) {
+        Node<T> current;
+        if (index <= size / 2) {
+            current = head;
+            for (int i = 0; i < index; i++) {
+                current = current.next;
+            }
+        } else {
+            current = tail;
+            for (int i = size - 1; i > index; i--) {
+                current = current.prev;
+            }
+        }
+        return current;
+    }
+
+    private void checkIndexInBounds(int index) {
+        if (index < 0 || index >= size) {
+            throw new IndexOutOfBoundsException("Incorrect index: " + index);
+        }
+    }
+
+    private void linkLast(T value) {
+        Node<T> newNode = new Node<>(null, value, null);
+        if (isEmpty()) {
+            tail = newNode;
+            head = newNode;
+        } else {
+            newNode.prev = tail;
+            tail.next = newNode;
+            tail = newNode;
+        }
+    }
+
+    private void linkBefore(T e, Node<T> before) {
```

Comment:
```suggestion
    private void linkBefore(T value, Node<T> before) {
```
Don’t  use abbreviations